### ISSUE TITLE
Is MS Bot Framework production ready?

### ISSUE THREAD

**Issue by McGern** [Microsoft/BotBuilder#2203]
Hi,

I know that the bot framework is in flux (i.e. being updated a lot) and there is a small "preview" written on the botframework website, but is it considered production ready? Is anyone using it for live work? I am currently using the webchat interface and am having a lot of issues with the service, mainly with directline (which I assume is the backbone for all the other interfaces as well?). The bot framework website mentions that is should be generally available by CY 2016 https://docs.botframework.com/en-us/faq/#how-long-will-the-bot-framework-be-in-preview-can-i-start-buildingshipping-products-based-on-a-preview-framework but don't the current state-of-play.

I am asking because we plunged into the world of bots using the MS Bot framework because we have the most experience with the MS stack and it also looked like a pretty flexible platform and I guess because I was banking on the Microsoft brand. The current project we are building is proving to be quite unstable, so we are on the verge of moving to a different platform - once that happens we are unlikely to invest more time in the MS framework.

In the fast paced world of bot development, there are new frameworks springing up seemingly every other week. I would love to continue to use the MS framework, but in reality I can't invest any more time if it doesn't work, and will need to investigate something that does. Are any other devs in a similar position?

I am not meaning to gripe, but I am currently in a hard place trying to justify what I believe is an excellent platform, and the reality of client demands.

Thanks,
Jarrod

**Comment by Andrea-Orimoto** [comment]
Hi Jarrod,
thanks a lot for your open and sincere feedback. 

The Bot Framework is still a preview technology and as such, we are still learning what developers are looking for and making frequent improvements. It is important to state that many companies are currently live or close to go live with their bots on our platform, for example:  [https://blogs.microsoft.com/firehose/2016/12/13/say-hello-to-the-new-skype-bots-from-expedia-com-ups-and-stubhub-among-others/](https://blogs.microsoft.com/firehose/2016/12/13/say-hello-to-the-new-skype-bots-from-expedia-com-ups-and-stubhub-among-others/)

We do provide help on our best capacity, even while not having a completely formalized support model as you would with technologies that are GA.

The whole bot industry is young, but we are confident Microsoft is best placed to support enterprise grade companies and their systems.

Having said that, would you be able to provide more details on the issues you have been facing? Perhaps we can give you more clarity on what we have been doing to solve them.

Thank you
Andrea

**Comment by McGern** [comment]
Hi Andrea, 

Thanks for getting back. Sorry this looks a big whinge, because I guess it is. I am quite enjoying the development side of the bot builder, and once I get my head around the dialogue stacking a bit more,  the bots should start rolling out. The main reason we went with the MS stack was because of the level of integration to other services and the potential speed getting a bot up and running - while other competing bots do some of this they left the web chat to the side "here's the api, get on with it"   

We are currently using the bot framework for web chat, so the services and technologies we are using,  provided under the Microsoft umbrella,  are:-
- bot builder (c# and nodejs sdk) 
- bot framework (connector) 
- directline to connect to web chat client
- web chat client (react components) 
- azure to host the bot and webchat

The main issues for me, and my thoughts, at the moment are (some of which have been explained away)
- constant directline 503 errors
   = don't know if it is is linked, but with these errors there is usually an accompanying xhttprequest cross-domain error
   = getting "couldn't send retry" timeouts with the web chat
- directline service is very slow
   = the conversation start is often a few seconds before a reply comes in, then another few seconds before conversationUpdate fires back a welcome message. This first call often fails entirely 
   = many of the polling "watermark" calls are shown as pending for a long time

My thoughts are that for an enterprise solution, there should be very few failures of service and latency when you are competing with the likes of Amazon and Facebook. Most other services I have tried (NLP or otherwise) have had quite nippy responses. I have read some of the answers from Dan Driscol regarding security and keeping order of conversation, and I won't pretend to understand all of it, but some of it seems overkill at the expense of speed - 200ms each way just for directline. 

- some of the services feel a bit bloated
   = using react + typescript + other supporting services to create a ~400kb minified web chat component seems like a lot of code for what, on the face of it, is a simple text based interface. 
   = the watermark sent by the web chat is very frequent. Don't know if this is a web chat thing or a directline thing. 
   = joining up all of the services is a bit daunting:- host the bot, connect to bot framework, connect to services, if LUIS as well get azure service key. You can't test the web chat with the local dev bot even though the emulator looks like it was made with the same code. 

Those are the main gripes, but as I mentioned before there a plenty of great things as well - if the latency and reliability are solved then it's a winner. 

Thanks for listening. 

Jarrod

**Comment by vovanb** [comment]
@Andrea-Orimoto 
I am interesting also when you will move to GA. In addition the response time on opened issue should be improved. We are waiting on some of items weeks, and sometime it is showstopper.
Thanks

**Comment by michaelvol** [comment]
We're also interested when is GA, and what is on the roadmap for the GA.
The bad handling of fast user inputs (https://github.com/Microsoft/BotBuilder/issues/2103, https://github.com/Microsoft/BotBuilder/issues/2119) is a serious issue for us. Not sure if/when it's to be addressed. 
Thanks!

**Comment by jferr** [comment]
Related question.  Is there a way to ask microsoft questions?  On the support page I see that stackoverflow is the recommended community support path but we  have questions that we'd like to ask microsoft that we don't want public.  Is there a microsoft email support address for the ms bot framework?

**Comment by upendra1588** [comment]
I am  facing **getting "couldn't send retry" timeouts with the web chat and emulator both**  is there any fix or know issue still?

**Comment by JasonSowers** [comment]
@upendra1588 can you please post a new issue for your question and post your code.  There is most likely something wrong with your bot itself if that is the error message you are getting.  We would like to help you but do not have enough information at the moment.

**Comment by mcsee** [comment]
WE have several bots in production and getting lots of 503s right now
We don't mind getting them but would like to have a status page to check like services from Google or Facebook 

https://azure.microsoft.com/en-us/status/ says everithing is ok
